rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,851 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// Author: Juliana Fajardini <jufajardini@oisf.net>
+
+//! PostgreSQL parser
+
+use super::parser::{self, PgsqlBEMessage, PgsqlFEMessage};
+use crate::applayer::{self, *};
+use crate::core::{self, AppProto, Flow, ALPROTO_UNKNOWN, IPPROTO_TCP};
+use nom;
+use std;
+use std::ffi::CString;
+
+static mut ALPROTO_PGSQL: AppProto = ALPROTO_UNKNOWN;
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialOrd, PartialEq, Debug)]
+pub enum PgsqlTransactionState {
+    Init = 0,
+    RequestReceived,
+    ResponseDone,
+}
+
+#[derive(Debug)]
+pub struct PgsqlTransaction {
+    pub tx_id: u64,
+    pub tx_state: PgsqlTransactionState,
+    pub requests: Vec<PgsqlFEMessage>,
+    pub responses: Vec<PgsqlBEMessage>,
+
+    de_state: Option<*mut core::DetectEngineState>,
+    events: *mut core::AppLayerDecoderEvents,
```

Comment:
Before merge to master I think this should be updated to be like the parsers after the very recent merge of putting these fields in the `tx_data` to remove some boilerplate.  Look at the DNS parser and see how we implement `State` and `Transaction` when can register some generic functions to handle a few things.

Reply:
Thanks, addressing that right now! :)